### PR TITLE
feat: add CLI

### PR DESCRIPTION
**Pull request by sith0008**
- fully implemented CLI tool
- wire up with communicator
- inherit from base request obejct
- added tests
- added `close() `method in communicators'
- [NEW] implemented `generateResponse` method
- [NEW] implemented `registerInterest` use case (spec says assume monitoring means user cannot perform other operations during monitoring interval, no need do multithreading)

TODO:
- to complete testing once server is up
- to review Datetime object (do we want to validate input or jus assume correct input? use custom datetime or java datetime? )

### CODE REVIEW

File: src/main/java/client/ServiceManager.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        Scanner scanner = new Scanner(System.in);
+        System.out.println("Please enter facility name: ");
+        String facilityName = scanner.nextLine();
+        System.out.println("Please enter the start date [DD/MM/YYYY]: ");
+        String startDate = scanner.nextLine();
+        System.out.println("Please enter the end date [DD/MM/YYYY]: ");
+        String endDate = scanner.nextLine();
+
+        List<Datetime> dates = new ArrayList<>();
+        dates.addAll(getListOfDates(startDate, endDate));
```
Comment (limxuanyu127) [inline src/main/java/client/ServiceManager.java]: Just wanted to highlight that server/communicator module need to convert it to LocalDateTime object for server to manipulate (import java.time.*;)

**Review by limxuanyu127 — APPROVED** [review]
looks gooood

**Review by xqlim1996 — APPROVED** [review]
Seems like no problems with merging, lets merge first and see if any potential problems arise

**Review by xqlim1996 — APPROVED** [review]
LGTM. thanks for adding close()

File: src/main/java/client/ServiceManager.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        String endDatetime = scanner.nextLine();
+
+        Request req = new BookFacilityRequest(
+                facilityName,
+                getDatetimeFromString(startDatetime),
+                getDatetimeFromString(endDatetime)
+        );
+        request(router, req);
+    }
+    public void offsetBooking() {
```
Comment (limxuanyu127) [inline src/main/java/client/ServiceManager.java:52]: Need to get facilname from user

File: src/main/java/client/ServiceManager.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                "1 = push back by 30 minutes \n" +
+                "2 = push back by 1 hour \n " +
+                "-1 = push forward by 30 minutes \n" +
+                "and so on. ");
+        int offset = Integer.parseInt(scanner.nextLine());
+
+        Request req = new OffsetBookingRequest(bookingID, offset);
+        request(router, req);
+    }
+    public void updateBooking() {
```
Comment (limxuanyu127) [inline src/main/java/client/ServiceManager.java]: Need to get facil name from user

File: src/main/java/client/ServiceManager.java (hunk 2 above)
Comment (sith0008) [reply src/main/java/client/ServiceManager.java:52]: oh booking ID is not enough?

**Comment by xqlim1996** [comment]
@sith0008 when you merge, extend Request for all the Request files and Add close() for the Server Communicator. Think the conflict is due to the line numbers ps

**Review by xqlim1996 — APPROVED** [review]
lgtm

**Review by limxuanyu127 — APPROVED** [review]

**Pull request merged**